ShapeIdeasReversePInvoke.cs

```diff
@@ -0,0 +1,584 @@
+        /// Returns the managed array.
+        /// </summary>
+        public T[]? ToManaged() => _managedArray;
+
+        /// <summary>
+        /// Frees native resources.
+        /// </summary>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.UnmanagedResources"/>
+        /// </remarks>
+        public void FreeNative()
+        {
+            Marshal.FreeCoTaskMem(_allocatedMemory);
+        }
+    }
+}
+
+public static unsafe partial class NativeImport
+{
+    public static partial TManaged CallFunc(
+        TManaged value,
+        in TManaged in_value,
+        ref TManaged ref_value,
+        out TManaged out_value);
+
+    // [return: MarshalUsing(ConstantElementCount = 30)]
+    public static unsafe partial TManaged[] CallFunc(
+        int[] value_blittable,
+        TManaged[] value,
+        [Out] TManaged[] out_byvalue,
+        in TManaged[] in_value,
+        // [MarshalUsing(ConstantElementCount = 10)]
+        ref TManaged[] ref_value,
+        // [MarshalUsing(ConstantElementCount = 20)]
+        out TManaged[] out_value);
+}
+
+#nullable disable
+
+public static unsafe partial class NativeImport
+{
+    public static partial TManaged CallFunc(
+        TManaged value,
+        in TManaged in_value,
+        ref TManaged ref_value,
+        out TManaged out_value)
+    {
+        Unsafe.SkipInit(out out_value);
+
+        Marshaller.ManagedToNative _byval_m_ = default;
+        TNative _byval_ = default;
+
+        Marshaller.ManagedToNative _in_m_ = default;
+        TNative _in_ = default;
+
+        Marshaller.Bidirectional _ref_m_ = default;
+        TNative _ref_ = default;
+
+        Marshaller.NativeToManaged _out_m_ = default;
+        TNative _out_ = default;
+
+        Marshaller.NativeToManaged _ret_m_ = default;
+        TNative _ret_ = default;
+
+        TManaged _ret_value_ = default;
+
+        try
+        {
+            // Marshal
+            _byval_m_.FromManaged(value);
+            fixed (void* _byval_dummy_ = &_byval_m_.GetPinnableReference())
+            {
+                _byval_ = _byval_m_.ToNative();
+
+                _in_m_.FromManaged(in_value);
+                fixed (void* _in_dummy_ = &_in_m_.GetPinnableReference())
+                {
+                    _in_ = _in_m_.ToNative();
+
+                    _ref_m_.FromManaged(ref_value);
+                    fixed (void* _ref_dummy_ = &_ref_m_.GetPinnableReference())
+                    {
+                        _ref_ = _ref_m_.ToNative();
+
+                        // Invoke
+                        _ret_ = __PInvoke(_byval_, &_in_, &_ref_, &_out_);
+                    }
+                }
+            }
+
+            // Unmarshal
+            _ref_m_.FromNative(_ref_);
+            ref_value = _ref_m_.ToManaged();
+
+            _out_m_.FromNative(_out_);
+            out_value = _out_m_.ToManaged();
+
+            _ret_m_.FromNative(_ret_);
+            _ret_value_ = _ret_m_.ToManaged();
```

Comment:
```suggestion
            _ref_m_.FromNative(_ref_);
            _out_m_.FromNative(_out_);
            _ret_m_.FromNative(_ret_);

            ref_value = _ref_m_.ToManaged();
            out_value = _out_m_.ToManaged();
            _ret_value_ = _ret_m_.ToManaged();
```
Should this be ordered like this to fix the memory leaks when exception are thrown during unmarshalling?

Reply:
This was conceptual to see the pairings and understand the example in a 1:1 manner. The source generation was what is shown now, but it now makes it more difficult to see how each case is handled. Not a big deal, but it was intentional.

Reply:
I think today, the source generation is actually more of what was done beforehand. We don't have two separate "unmarshalling" stages. That's something we'll need to add (which I think is a good thing).